Library/Homebrew/dev-cmd/linkage.rb

```diff
@@ -16,32 +16,30 @@
 
 module Homebrew
   def linkage
-    found_broken_dylibs = false
     ARGV.kegs.each do |keg|
       ohai "Checking #{keg.name} linkage" if ARGV.kegs.size > 1
       result = LinkageChecker.new(keg)
       if ARGV.include?("--test")
         result.display_test_output
+        Homebrew.failed = true if result.broken_dylibs.any? || result.undeclared_deps.any?
```

Comment:
Updated. I also amend the PR with tweaking the logic of undeclared_deps to handle some edge cases.